presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ExtractSpatialJoins.java

```diff
@@ -561,55 +561,67 @@ private static int checkAlignment(JoinNode joinNode, Set<VariableReferenceExpres
         }
 
-        projections.put(variable, castToRowExpression(expression));
+        projections.put(variable, expression);
         return new ProjectNode(context.getIdAllocator().getNextId(), node, projections.build());
     }
 
-    private static PlanNode addPartitioningNodes(Context context, PlanNode node, VariableReferenceExpression partitionVariable, KdbTree kdbTree, Expression geometry, Optional<Expression> radius)
+    private static PlanNode addPartitioningNodes(Context context, FunctionManager functionManager, PlanNode node, VariableReferenceExpression partitionVariable, KdbTree kdbTree, RowExpression geometry, Optional<RowExpression> radius)
     {
         Assignments.Builder projections = Assignments.builder();
         for (VariableReferenceExpression outputVariable : node.getOutputVariables()) {
-            projections.put(identityAsSymbolReference(outputVariable));
+            projections.put(outputVariable, outputVariable);
         }
 
-        ImmutableList.Builder<Expression> partitioningArguments = ImmutableList.<Expression>builder()
-                .add(new Cast(new StringLiteral(KdbTreeUtils.toJson(kdbTree)), KDB_TREE_TYPENAME))
-                .add(geometry);
-        radius.map(partitioningArguments::add);
+        FunctionHandle castFunctionHandle = functionManager.lookupCast(CAST, VARCHAR.getTypeSignature(), KDB_TREE.getTypeSignature());
 
-        FunctionCall partitioningFunction = new FunctionCall(QualifiedName.of("spatial_partitions"), partitioningArguments.build());
-        VariableReferenceExpression partitionsVariable = context.getVariableAllocator().newVariable(partitioningFunction, new ArrayType(INTEGER));
-        projections.put(partitionsVariable, castToRowExpression(partitioningFunction));
+        ImmutableList.Builder<RowExpression> partitioningArgumentsBuilder = ImmutableList.<RowExpression>builder()
+                .add(new CallExpression(CAST.name(), castFunctionHandle, KDB_TREE, ImmutableList.of(new ConstantExpression(utf8Slice(KdbTreeUtils.toJson(kdbTree)), VARCHAR))))
+                .add(geometry);
+        radius.map(partitioningArgumentsBuilder::add);
+        ImmutableList<RowExpression> partitioningArguments = partitioningArgumentsBuilder.build();
+
+        QualifiedName spatialPartitions = QualifiedName.of("spatial_partitions");
+        FunctionHandle functionHandle = functionManager.lookupFunction(
+                spatialPartitions.toString(),
+                fromTypes(partitioningArguments.stream()
+                        .map(RowExpression::getType).collect(toImmutableList())));
+
+        CallExpression partitioningFunction = new CallExpression(spatialPartitions.toString(), functionHandle, new ArrayType(INTEGER), partitioningArguments);
+        VariableReferenceExpression partitionsVariable = context.getVariableAllocator().newVariable(partitioningFunction);
+        projections.put(partitionsVariable, partitioningFunction);
```

Comment:
Function resolution here is another thing I'm unsure I did correctly. Let me know if the way I resolve the cast call is correct etc.